### ISSUE TITLE
Strange build dependencies from Tofino backend

### ISSUE THREAD

**Issue by asl** [p4lang/p4c#5013]
Building e.g. a `frontend` target actually builds Tofino backend, which does not look correct to me.

**Comment by fruffy** [comment]
If I'd wager a guess it might be because of the Tofino IR sources: https://github.com/p4lang/p4c/blob/main/backends/tofino/CMakeLists.txt#L219 which need to be compile with the ir-generated blob.

**Comment by asl** [comment]
> https://github.com/p4lang/p4c/blob/main/backends/tofino/CMakeLists.txt#L219 which need to be compile with the ir-generated blob.

But looks like these are not entirely frontend...

**Comment by fruffy** [comment]
>But looks like these are not entirely frontend...

Yeah I also think they could be cleaned up. There includes are quite messy and there are some circular dependencies.

**Comment by ChrisDodd** [comment]
The way the code is set up, it builds IR Visitor infrastructure as part of the frontend, supporting all IR types that are declared in any backend that is enabled in the build as part of the frontend.  That ends up pulling a few backend files (those related to the IR types) into the frontend, and is required due C++ ODR requirements (otherwise you would get undefined behavior by linking code that doesn't agree on the vtable layout for classes like IR::Node)

**Comment by asl** [comment]
> The way the code is set up, it builds IR Visitor infrastructure as part of the frontend, supporting all IR types that are declared in any backend that is enabled in the build as part of the frontend. That ends up pulling a few backend files (those related to the IR types) into the frontend, and is required due C++ ODR requirements (otherwise you would get undefined behavior by linking code that doesn't agree on the vtable layout for classes like IR::Node).
> 
> To not include this backend code, you need to use `-DENABLE_TOFINO=OFF` to cmake and rebuild everything.

But are files like `bf-p4c/midend/path_linearizer.cpp` or `bf-utils/src/dynamic_hash/dynamic_hash.c` `bf-utils/src/dynamic_hash/bfn_hash_algorithm.c` constitute IR Visitor infrastructure?

**Comment by ChrisDodd** [comment]
> But are files like `bf-p4c/midend/path_linearizer.cpp` or `bf-utils/src/dynamic_hash/dynamic_hash.c` `bf-utils/src/dynamic_hash/bfn_hash_algorithm.c` constitute IR Visitor infrastructure?

No, but disentangling them is likely to require splitting some source files to only include the minimum to get things to build while not including other things that pull in more dependencies.
